cognite/extractorutils/uploader.py

```diff
@@ -809,3 +810,115 @@ def __len__(self) -> int:
+    """
+    Upload queue for files
+
+    Args:
+        cdf_client: Cognite Data Fusion client to use
+        post_upload_function: A function that will be called after each upload. The function will be given one argument:
+            A list of the events that were uploaded.
+        max_queue_size: Maximum size of upload queue. Defaults to no max size.
+        max_upload_interval: Automatically trigger an upload each m seconds when run as a thread (use start/stop
+            methods).
+        trigger_log_level: Log level to log upload triggers to.
+        thread_name: Thread name of uploader thread.
+    """
+
+    def __init__(
+        self,
+        cdf_client: CogniteClient,
+        post_upload_function: Optional[Callable[[List[Event]], None]] = None,
+        max_queue_size: Optional[int] = None,
+        max_upload_interval: Optional[int] = None,
+        trigger_log_level: str = "DEBUG",
+        thread_name: Optional[str] = None,
+        overwrite_existing: bool = False,
+    ):
+        # Super sets post_upload and threshold
+        super().__init__(
+            cdf_client, post_upload_function, max_queue_size, max_upload_interval, trigger_log_level, thread_name
+        )
+
+        self.upload_queue: List[Tuple[FileMetadata, Union[str, PathLike]]] = []
+        self.overwrite_existing = overwrite_existing
+
+    def add_to_upload_queue(self, file_meta: FileMetadata, file_name: Union[str, PathLike] = None) -> None:
+        """
+        Add file to upload queue. The queue will be uploaded if the queue size is larger than the threshold
+        specified in the __init__.
+
+        Args:
+            file_meta: File metadata-object
+            file_name: Path to file to be uploaded.
+                If none, the file object will still be created, but no data is uploaded
+        """
+        with self.lock:
+            self.upload_queue.append((file_meta, file_name))
+            self.upload_queue_size += 1
+
+            self._check_triggers()
+
+    def upload(self) -> None:
+        """
+        Trigger an upload of the queue, clears queue afterwards
+        """
+        if len(self.upload_queue) == 0:
+            return
+
+        with self.lock:
+            self._upload_batch()
+            try:
+                self._post_upload(self.upload_queue)
+            except Exception as e:
+                self.logger.error("Error in upload callback: %s", str(e))
+            self.upload_queue.clear()
+            self.upload_queue_size = 0
+
+    @retry(
+        exceptions=(CogniteAPIError, ConnectionError),
+        tries=RETRIES,
+        delay=RETRY_DELAY,
+        max_delay=RETRY_MAX_DELAY,
+        backoff=RETRY_BACKOFF_FACTOR,
+    )
+    def _upload_batch(self):
+        for i, (file_meta, file_name) in enumerate(self.upload_queue):
+
+            # Upload file
+            file_meta = self.cdf_client.files.upload(file_name, **file_meta.dump(), overwrite=self.overwrite_existing)
```

Comment:
For all other resource types, the SDK is doing some chunking and parallelization of requests, but since the API doesn't support multiple files in a single upload, it doesn't do it for files. 

I think it would make sense for us to add some parallelization to this since we're batching up files for uploading, something like
```python
with ThreadPoolExecutor(self.cdf_client.config.max_workers) as pool:
    for i, (file_meta, file_name) in enumerate(self.upload_queue):
        pool.submit(fn=self._upload_single, args=(file_name, file_meta))
```
Maybe also move the retry logic to the individual uploads to avoid redoing all the uploading on a single failure.

Reply:
This is a good idea. I have updated this function to reflect this.